File: src/backend/utils/date_utils.py
"""
Date and time utility functions for the Freight Price Movement Agent.

This module provides comprehensive date and time manipulation capabilities including
parsing, formatting, arithmetic operations, period calculations, and timezone handling.
These utilities support the time period selection and analysis features of the application.
"""

import datetime
from dateutil import parser
from dateutil import relativedelta
import pytz
from typing import List, Optional, Tuple, Union

# Default timezone and date format constants
DEFAULT_TIMEZONE = pytz.UTC
DATE_FORMAT = "%Y-%m-%d"
DATETIME_FORMAT = "%Y-%m-%d %H:%M:%S"
ISO_FORMAT = "%Y-%m-%dT%H:%M:%S%z"


def parse_date(date_string: str, format_string: Optional[str] = None, 
               timezone: Optional[datetime.tzinfo] = None) -> datetime.datetime:
    """
    Parse a date string into a datetime object.
    
    Args:
        date_string: String representation of a date/datetime
        format_string: Optional format string for parsing (if None, uses flexible parsing)
        timezone: Optional timezone to assign to the parsed datetime
        
    Returns:
        A datetime object representing the parsed date
        
    Raises:
        ValueError: If the date string cannot be parsed
    """
    try:
        if format_string:
            # Parse using the specified format
            parsed_date = datetime.datetime.strptime(date_string, format_string)
        else:
            # Use flexible parsing
            parsed_date = parser.parse(date_string)
        
        # Handle timezone
        if timezone:
            if parsed_date.tzinfo is None:
                # Assign timezone to naive datetime
                parsed_date = parsed_date.replace(tzinfo=timezone)
            else:
                # Convert to specified timezone
                parsed_date = parsed_date.astimezone(timezone)
        elif parsed_date.tzinfo is None:
            # If no timezone specified and parsed date is naive, use default
            parsed_date = parsed_date.replace(tzinfo=DEFAULT_TIMEZONE)
            
        return parsed_date
    except (ValueError, TypeError) as e:
        raise ValueError(f"Failed to parse date string '{date_string}': {str(e)}")


def format_date(dt: datetime.datetime, format_string: Optional[str] = None) -> str:
    """
    Format a datetime object as a date string.
    
    Args:
        dt: The datetime object to format
        format_string: Optional format string (defaults to DATE_FORMAT)
        
    Returns:
        Formatted date string
        
    Raises:
        ValueError: If the datetime cannot be formatted
    """
    try:
        if format_string is None:
            format_string = DATE_FORMAT
        return dt.strftime(format_string)
    except (ValueError, TypeError) as e:
        raise ValueError(f"Failed to format datetime {dt}: {str(e)}")


def format_datetime(dt: datetime.datetime, format_string: Optional[str] = None) -> str:
    """
    Format a datetime object as a string with time component.
    
    Args:
        dt: The datetime object to format
        format_string: Optional format string (defaults to DATETIME_FORMAT)
        
    Returns:
        Formatted datetime string
        
    Raises:
        ValueError: If the datetime cannot be formatted
    """
    try:
        if format_string is None:
            format_string = DATETIME_FORMAT
        return dt.strftime(format_string)
    except (ValueError, TypeError) as e:
        raise ValueError(f"Failed to format datetime {dt}: {str(e)}")


def to_iso_format(dt: datetime.datetime) -> str:
    """
    Convert a datetime object to ISO 8601 format string.
    
    Args:
        dt: The datetime object to format
        
    Returns:
        ISO formatted datetime string
        
    Raises:
        ValueError: If the datetime cannot be formatted
    """
    try:
        # Ensure datetime has timezone information
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=DEFAULT_TIMEZONE)
        return dt.isoformat()
    except (ValueError, TypeError) as e:
        raise ValueError(f"Failed to convert datetime to ISO format: {str(e)}")


def now(timezone: Optional[datetime.tzinfo] = None) -> datetime.datetime:
    """
    Get the current datetime with timezone information.
    
    Args:
        timezone: Optional timezone (defaults to DEFAULT_TIMEZONE)
        
    Returns:
        Current datetime with timezone
    """
    current = datetime.datetime.now()
    if timezone is None:
        timezone = DEFAULT_TIMEZONE
    
    if current.tzinfo is None:
        current = current.replace(tzinfo=timezone)
    else:
        current = current.astimezone(timezone)
    
    return current


def today(timezone: Optional[datetime.tzinfo] = None) -> datetime.datetime:
    """
    Get the current date (without time component).
    
    Args:
        timezone: Optional timezone (defaults to DEFAULT_TIMEZONE)
        
    Returns:
        Current date at midnight
    """
    current = now(timezone)
    return current.replace(hour=0, minute=0, second=0, microsecond=0)


def convert_timezone(dt: datetime.datetime, target_timezone: datetime.tzinfo) -> datetime.datetime:
    """
    Convert a datetime from one timezone to another.
    
    Args:
        dt: The datetime to convert
        target_timezone: The target timezone
        
    Returns:
        Datetime in target timezone
        
    Raises:
        ValueError: If the datetime cannot be converted
    """
    try:
        # Ensure datetime has timezone information
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=DEFAULT_TIMEZONE)
        
        # Convert to target timezone
        return dt.astimezone(target_timezone)
    except (ValueError, TypeError) as e:
        raise ValueError(f"Failed to convert timezone: {str(e)}")


def normalize_to_utc(dt: datetime.datetime) -> datetime.datetime:
    """
    Normalize a datetime to UTC timezone.
    
    Args:
        dt: The datetime to normalize
        
    Returns:
        Datetime in UTC timezone
    """
    return convert_timezone(dt, pytz.UTC)


def add_days(dt: datetime.datetime, days: int) -> datetime.datetime:
    """
    Add a specified number of days to a datetime.
    
    Args:
        dt: The base datetime
        days: Number of days to add (can be negative)
        
    Returns:
        Datetime with days added
        
    Raises:
        ValueError: If the operation fails
    """
    try:
        return dt + datetime.timedelta(days=days)
    except (ValueError, TypeError, OverflowError) as e:
        raise ValueError(f"Failed to add {days} days to datetime: {str(e)}")


def add_weeks(dt: datetime.datetime, weeks: int) -> datetime.datetime:
    """
    Add a specified number of weeks to a datetime.
    
    Args:
        dt: The base datetime
        weeks: Number of weeks to add (can be negative)
        
    Returns:
        Datetime with weeks added
        
    Raises:
        ValueError: If the operation fails
    """
    try:
        return dt + datetime.timedelta(weeks=weeks)
    except (ValueError, TypeError, OverflowError) as e:
        raise ValueError(f"Failed to add {weeks} weeks to datetime: {str(e)}")


def add_months(dt: datetime.datetime, months: int) -> datetime.datetime:
    """
    Add a specified number of months to a datetime.
    
    Args:
        dt: The base datetime
        months: Number of months to add (can be negative)
        
    Returns:
        Datetime with months added
        
    Raises:
        ValueError: If the operation fails
    """
    try:
        return dt + relativedelta.relativedelta(months=months)
    except (ValueError, TypeError) as e:
        raise ValueError(f"Failed to add {months} months to datetime: {str(e)}")


def add_quarters(dt: datetime.datetime, quarters: int) -> datetime.datetime:
    """
    Add a specified number of quarters (3-month periods) to a datetime.
    
    Args:
        dt: The base datetime
        quarters: Number of quarters to add (can be negative)
        
    Returns:
        Datetime with quarters added
    """
    return add_months(dt, quarters * 3)


def date_diff_days(start_date: datetime.datetime, end_date: datetime.datetime) -> int:
    """
    Calculate the difference between two dates in days.
    
    Args:
        start_date: The earlier datetime
        end_date: The later datetime
        
    Returns:
        Number of days between dates
        
    Raises:
        ValueError: If the calculation fails
    """
    try:
        delta = end_date - start_date
        return abs(delta.days)
    except (ValueError, TypeError, OverflowError) as e:
        raise ValueError(f"Failed to calculate days between dates: {str(e)}")


def date_diff_weeks(start_date: datetime.datetime, end_date: datetime.datetime) -> float:
    """
    Calculate the difference between two dates in weeks.
    
    Args:
        start_date: The earlier datetime
        end_date: The later datetime
        
    Returns:
        Number of weeks between dates (rounded to 2 decimal places)
    """
    days = date_diff_days(start_date, end_date)
    weeks = days / 7
    return round(weeks, 2)


def date_diff_months(start_date: datetime.datetime, end_date: datetime.datetime) -> int:
    """
    Calculate the approximate difference between two dates in months.
    
    Args:
        start_date: The earlier datetime
        end_date: The later datetime
        
    Returns:
        Number of months between dates
        
    Raises:
        ValueError: If the calculation fails
    """
    try:
        # Normalize dates to same timezone if they have timezone info
        if start_date.tzinfo and end_date.tzinfo:
            start_date = normalize_to_utc(start_date)
            end_date = normalize_to_utc(end_date)
        
        # Ensure start_date is before end_date
        if start_date > end_date:
            start_date, end_date = end_date, start_date
        
        # Calculate using relativedelta
        delta = relativedelta.relativedelta(end_date, start_date)
        months = delta.months + delta.years * 12
        
        return months
    except (ValueError, TypeError) as e:
        raise ValueError(f"Failed to calculate months between dates: {str(e)}")


def date_diff_quarters(start_date: datetime.datetime, end_date: datetime.datetime) -> float:
    """
    Calculate the approximate difference between two dates in quarters.
    
    Args:
        start_date: The earlier datetime
        end_date: The later datetime
        
    Returns:
        Number of quarters between dates (rounded to 2 decimal places)
    """
    months = date_diff_months(start_date, end_date)
    quarters = months / 3
    return round(quarters, 2)


def get_start_of_day(dt: datetime.datetime) -> datetime.datetime:
    """
    Get the start of the day (midnight) for a given datetime.
    
    Args:
        dt: The input datetime
        
    Returns:
        Datetime at start of day
    """
    return dt.replace(hour=0, minute=0, second=0, microsecond=0)


def get_end_of_day(dt: datetime.datetime) -> datetime.datetime:
    """
    Get the end of the day (23:59:59) for a given datetime.
    
    Args:
        dt: The input datetime
        
    Returns:
        Datetime at end of day
    """
    return dt.replace(hour=23, minute=59, second=59, microsecond=999999)


def get_start_of_week(dt: datetime.datetime) -> datetime.datetime:
    """
    Get the start of the week (Monday) for a given datetime.
    
    Args:
        dt: The input datetime
        
    Returns:
        Datetime at start of week
    """
    # Subtract days to get to Monday (weekday() returns 0 for Monday)
    days_to_subtract = dt.weekday()
    start_of_week = dt - datetime.timedelta(days=days_to_subtract)
    return get_start_of_day(start_of_week)


def get_end_of_week(dt: datetime.datetime) -> datetime.datetime:
    """
    Get the end of the week (Sunday) for a given datetime.
    
    Args:
        dt: The input datetime
        
    Returns:
        Datetime at end of week
    """
    # Add days to get to Sunday (6 - weekday() gives days to add)
    days_to_add = 6 - dt.weekday()
    end_of_week = dt + datetime.timedelta(days=days_to_add)
    return get_end_of_day(end_of_week)


def get_start_of_month(dt: datetime.datetime) -> datetime.datetime:
    """
    Get the start of the month for a given datetime.
    
    Args:
        dt: The input datetime
        
    Returns:
        Datetime at start of month
    """
    start_of_month = dt.replace(day=1)
    return get_start_of_day(start_of_month)


def get_end_of_month(dt: datetime.datetime) -> datetime.datetime:
    """
    Get the end of the month for a given datetime.
    
    Args:
        dt: The input datetime
        
    Returns:
        Datetime at end of month
    """
    # Get the first day of next month, then subtract one day
    next_month = dt.replace(day=1) + relativedelta.relativedelta(months=1)
    end_of_month = next_month - datetime.timedelta(days=1)
    return get_end_of_day(end_of_month)


def get_start_of_quarter(dt: datetime.datetime) -> datetime.datetime:
    """
    Get the start of the quarter for a given datetime.
    
    Args:
        dt: The input datetime
        
    Returns:
        Datetime at start of quarter
    """
    # Calculate the first month of the quarter (1, 4, 7, or 10)
    quarter_month = (dt.month - 1) // 3 * 3 + 1
    start_of_quarter = dt.replace(month=quarter_month, day=1)
    return get_start_of_day(start_of_quarter)


def get_end_of_quarter(dt: datetime.datetime) -> datetime.datetime:
    """
    Get the end of the quarter for a given datetime.
    
    Args:
        dt: The input datetime
        
    Returns:
        Datetime at end of quarter
    """
    # Get the start of the quarter, add 3 months, subtract one day
    start_of_next_quarter = add_months(get_start_of_quarter(dt), 3)
    end_of_quarter = start_of_next_quarter - datetime.timedelta(days=1)
    return get_end_of_day(end_of_quarter)


def generate_date_periods(
    start_date: datetime.datetime,
    end_date: datetime.datetime,
    granularity: str,
    custom_interval_days: Optional[int] = None
) -> List[Tuple[datetime.datetime, datetime.datetime]]:
    """
    Generate a list of date periods based on the specified granularity.
    
    Args:
        start_date: The start datetime of the overall period
        end_date: The end datetime of the overall period
        granularity: The granularity of periods ('daily', 'weekly', 'monthly', 
                     'quarterly', or 'custom')
        custom_interval_days: Number of days in each period when granularity is 'custom'
        
    Returns:
        List of tuples containing (period_start, period_end) datetimes
        
    Raises:
        ValueError: If dates are invalid or granularity is not supported
    """
    if start_date > end_date:
        raise ValueError("Start date must be before end date")
    
    periods = []
    current_start = start_date
    
    if granularity == 'daily':
        while current_start < end_date:
            current_end = get_end_of_day(current_start)
            if current_end > end_date:
                current_end = end_date
            periods.append((current_start, current_end))
            current_start = add_days(get_start_of_day(current_start), 1)
    
    elif granularity == 'weekly':
        # Adjust to start of week
        current_start = get_start_of_week(current_start)
        while current_start < end_date:
            current_end = get_end_of_week(current_start)
            if current_end > end_date:
                current_end = end_date
            periods.append((current_start, current_end))
            current_start = add_days(current_start, 7)
    
    elif granularity == 'monthly':
        # Adjust to start of month
        current_start = get_start_of_month(current_start)
        while current_start < end_date:
            current_end = get_end_of_month(current_start)
            if current_end > end_date:
                current_end = end_date
            periods.append((current_start, current_end))
            current_start = add_months(current_start, 1)
    
    elif granularity == 'quarterly':
        # Adjust to start of quarter
        current_start = get_start_of_quarter(current_start)
        while current_start < end_date:
            current_end = get_end_of_quarter(current_start)
            if current_end > end_date:
                current_end = end_date
            periods.append((current_start, current_end))
            current_start = add_months(current_start, 3)
    
    elif granularity == 'custom':
        if custom_interval_days is None or custom_interval_days <= 0:
            raise ValueError("Custom interval days must be a positive integer")
        
        while current_start < end_date:
            current_end = add_days(current_start, custom_interval_days - 1)
            current_end = get_end_of_day(current_end)
            if current_end > end_date:
                current_end = end_date
            periods.append((current_start, current_end))
            current_start = add_days(get_start_of_day(current_start), custom_interval_days)
    
    else:
        raise ValueError(f"Unsupported granularity: {granularity}")
    
    return periods


def is_same_day(dt1: datetime.datetime, dt2: datetime.datetime) -> bool:
    """
    Check if two datetimes fall on the same day.
    
    Args:
        dt1: First datetime
        dt2: Second datetime
        
    Returns:
        True if same day, False otherwise
    """
    # Normalize to same timezone if they have timezone info
    if dt1.tzinfo and dt2.tzinfo:
        dt1 = normalize_to_utc(dt1)
        dt2 = normalize_to_utc(dt2)
    
    return (dt1.year == dt2.year and 
            dt1.month == dt2.month and 
            dt1.day == dt2.day)


def is_future_date(dt: datetime.datetime, reference_date: Optional[datetime.datetime] = None) -> bool:
    """
    Check if a date is in the future.
    
    Args:
        dt: The datetime to check
        reference_date: The reference datetime (defaults to current datetime)
        
    Returns:
        True if future date, False otherwise
    """
    if reference_date is None:
        reference_date = now()
    
    # Normalize to same timezone if they have timezone info
    if dt.tzinfo and reference_date.tzinfo:
        dt = normalize_to_utc(dt)
        reference_date = normalize_to_utc(reference_date)
    
    return dt > reference_date


def is_past_date(dt: datetime.datetime, reference_date: Optional[datetime.datetime] = None) -> bool:
    """
    Check if a date is in the past.
    
    Args:
        dt: The datetime to check
        reference_date: The reference datetime (defaults to current datetime)
        
    Returns:
        True if past date, False otherwise
    """
    if reference_date is None:
        reference_date = now()
    
    # Normalize to same timezone if they have timezone info
    if dt.tzinfo and reference_date.tzinfo:
        dt = normalize_to_utc(dt)
        reference_date = normalize_to_utc(reference_date)
    
    return dt < reference_date


def get_date_range_description(
    start_date: datetime.datetime,
    end_date: datetime.datetime,
    format_string: Optional[str] = None
) -> str:
    """
    Generate a human-readable description of a date range.
    
    Args:
        start_date: The start datetime
        end_date: The end datetime
        format_string: Optional format string for dates
        
    Returns:
        Human-readable date range description
        
    Raises:
        ValueError: If the formatting fails
    """
    try:
        start_str = format_date(start_date, format_string)
        end_str = format_date(end_date, format_string)
        return f"{start_str} to {end_str}"
    except ValueError as e:
        raise ValueError(f"Failed to create date range description: {str(e)}")

File: src/web/src/utils/date-utils.ts
/**
 * Utility functions for date manipulation and formatting in the Freight Price Movement Agent
 * Provides centralized date-related helper functions to ensure consistent date handling
 * across the application, particularly for time period selection and analysis.
 */

import dayjs from "dayjs";
import utc from "dayjs/plugin/utc"; // v1.11.7
import timezone from "dayjs/plugin/timezone"; // v1.11.7
import customParseFormat from "dayjs/plugin/customParseFormat"; // v1.11.7
import isSameOrBefore from "dayjs/plugin/isSameOrBefore"; // v1.11.7
import isSameOrAfter from "dayjs/plugin/isSameOrAfter"; // v1.11.7
import { DATE_FORMATS, DEFAULT_DATE_FORMAT } from '../config/constants';
import { TimeGranularity } from '../types';

// Extend dayjs with plugins
dayjs.extend(utc);
dayjs.extend(timezone);
dayjs.extend(customParseFormat);
dayjs.extend(isSameOrBefore);
dayjs.extend(isSameOrAfter);

/**
 * Formats a date string or Date object according to the specified format
 * 
 * @param date - Date string or Date object to format
 * @param format - Format string (defaults to application default format)
 * @returns Formatted date string
 */
export const formatDate = (date: string | Date, format: string = DEFAULT_DATE_FORMAT): string => {
  if (!date) {
    return '';
  }
  
  return dayjs(date).format(format);
};

/**
 * Parses a date string in the specified format into a Date object
 * 
 * @param dateString - Date string to parse
 * @param format - Format of the date string (defaults to application default format)
 * @returns Parsed Date object or null if invalid
 */
export const parseDate = (dateString: string, format: string = DEFAULT_DATE_FORMAT): Date | null => {
  if (!dateString) {
    return null;
  }
  
  const parsed = dayjs(dateString, format);
  
  return parsed.isValid() ? parsed.toDate() : null;
};

/**
 * Checks if a date string is valid according to the specified format
 * 
 * @param dateString - Date string to validate
 * @param format - Format to validate against (defaults to application default format)
 * @returns True if the date is valid, false otherwise
 */
export const isValidDate = (dateString: string, format: string = DEFAULT_DATE_FORMAT): boolean => {
  if (!dateString) {
    return false;
  }
  
  return dayjs(dateString, format).isValid();
};

/**
 * Formats a date for API requests using the API_FORMAT
 * 
 * @param date - Date string or Date object to format
 * @returns Date formatted for API consumption
 */
export const formatDateForAPI = (date: string | Date): string => {
  if (!date) {
    return '';
  }
  
  return dayjs(date).format(DATE_FORMATS.API_FORMAT);
};

/**
 * Formats a date received from the API to the application's display format
 * 
 * @param dateString - Date string from API (in API_FORMAT)
 * @returns Date formatted for display
 */
export const formatDateFromAPI = (dateString: string): string => {
  if (!dateString) {
    return '';
  }
  
  return dayjs(dateString, DATE_FORMATS.API_FORMAT).format(DEFAULT_DATE_FORMAT);
};

/**
 * Generates start and end dates for a predefined time period based on granularity
 * 
 * @param granularity - Time granularity (daily, weekly, monthly)
 * @param periods - Number of periods to go back
 * @returns Object with startDate and endDate strings in YYYY-MM-DD format
 */
export const getDateRangeForGranularity = (
  granularity: TimeGranularity,
  periods: number = 1
): { startDate: string; endDate: string } => {
  const endDate = dayjs();
  let startDate;

  switch (granularity) {
    case TimeGranularity.DAILY:
      startDate = endDate.subtract(periods, 'day');
      break;
    case TimeGranularity.WEEKLY:
      startDate = endDate.subtract(periods, 'week');
      break;
    case TimeGranularity.MONTHLY:
      startDate = endDate.subtract(periods, 'month');
      break;
    default:
      startDate = endDate.subtract(periods, 'day');
  }

  return {
    startDate: startDate.format(DATE_FORMATS.YYYY_MM_DD),
    endDate: endDate.format(DATE_FORMATS.YYYY_MM_DD),
  };
};

/**
 * Calculates the difference between two dates in the specified unit
 * 
 * @param startDate - Start date string or Date object
 * @param endDate - End date string or Date object
 * @param unit - Unit for calculation (day, week, month, etc.)
 * @returns Difference between dates in the specified unit
 */
export const calculateDateDifference = (
  startDate: string | Date,
  endDate: string | Date,
  unit: string = 'day'
): number => {
  const start = dayjs(startDate);
  const end = dayjs(endDate);
  
  return Math.abs(end.diff(start, unit as any));
};

/**
 * Adds a specified amount of time to a date
 * 
 * @param date - Date string or Date object
 * @param amount - Amount to add
 * @param unit - Unit of time to add (day, week, month, etc.)
 * @returns New date after addition, formatted as YYYY-MM-DD
 */
export const addToDate = (
  date: string | Date,
  amount: number,
  unit: string = 'day'
): string => {
  return dayjs(date).add(amount, unit as any).format(DATE_FORMATS.YYYY_MM_DD);
};

/**
 * Subtracts a specified amount of time from a date
 * 
 * @param date - Date string or Date object
 * @param amount - Amount to subtract
 * @param unit - Unit of time to subtract (day, week, month, etc.)
 * @returns New date after subtraction, formatted as YYYY-MM-DD
 */
export const subtractFromDate = (
  date: string | Date,
  amount: number,
  unit: string = 'day'
): string => {
  return dayjs(date).subtract(amount, unit as any).format(DATE_FORMATS.YYYY_MM_DD);
};

/**
 * Checks if a date is before another date
 * 
 * @param date - Date to check
 * @param compareDate - Date to compare against
 * @returns True if date is before compareDate, false otherwise
 */
export const isDateBefore = (date: string | Date, compareDate: string | Date): boolean => {
  return dayjs(date).isBefore(dayjs(compareDate));
};

/**
 * Checks if a date is after another date
 * 
 * @param date - Date to check
 * @param compareDate - Date to compare against
 * @returns True if date is after compareDate, false otherwise
 */
export const isDateAfter = (date: string | Date, compareDate: string | Date): boolean => {
  return dayjs(date).isAfter(dayjs(compareDate));
};

/**
 * Checks if a date is between two other dates (inclusive)
 * 
 * @param date - Date to check
 * @param startDate - Start of date range
 * @param endDate - End of date range
 * @returns True if date is between startDate and endDate (inclusive), false otherwise
 */
export const isDateBetween = (
  date: string | Date,
  startDate: string | Date,
  endDate: string | Date
): boolean => {
  const d = dayjs(date);
  return d.isSameOrAfter(dayjs(startDate)) && d.isSameOrBefore(dayjs(endDate));
};

/**
 * Gets the start of a period (day, week, month, etc.) for a given date
 * 
 * @param date - Date string or Date object
 * @param unit - Unit of time (day, week, month, etc.)
 * @returns Start of period date formatted as YYYY-MM-DD
 */
export const getStartOfPeriod = (date: string | Date, unit: string = 'day'): string => {
  return dayjs(date).startOf(unit as any).format(DATE_FORMATS.YYYY_MM_DD);
};

/**
 * Gets the end of a period (day, week, month, etc.) for a given date
 * 
 * @param date - Date string or Date object
 * @param unit - Unit of time (day, week, month, etc.)
 * @returns End of period date formatted as YYYY-MM-DD
 */
export const getEndOfPeriod = (date: string | Date, unit: string = 'day'): string => {
  return dayjs(date).endOf(unit as any).format(DATE_FORMATS.YYYY_MM_DD);
};

/**
 * Generates an array of date ranges based on start date, end date, and granularity
 * 
 * @param startDate - Start date string or Date object
 * @param endDate - End date string or Date object
 * @param granularity - Time granularity (daily, weekly, monthly, custom)
 * @param customInterval - Custom interval in days (used only when granularity is CUSTOM)
 * @returns Array of date range objects { start, end, label }
 */
export const generateDateRanges = (
  startDate: string | Date,
  endDate: string | Date,
  granularity: TimeGranularity,
  customInterval?: number
): Array<{ start: string; end: string; label: string }> => {
  const start = dayjs(startDate);
  const end = dayjs(endDate);
  const ranges: Array<{ start: string; end: string; label: string }> = [];
  
  let unit: string;
  let interval: number;
  
  // Determine unit and interval based on granularity
  switch (granularity) {
    case TimeGranularity.DAILY:
      unit = 'day';
      interval = 1;
      break;
    case TimeGranularity.WEEKLY:
      unit = 'week';
      interval = 1;
      break;
    case TimeGranularity.MONTHLY:
      unit = 'month';
      interval = 1;
      break;
    case TimeGranularity.CUSTOM:
      unit = 'day';
      interval = customInterval || 1;
      break;
    default:
      unit = 'day';
      interval = 1;
  }
  
  // Generate date ranges
  let currentStart = start;
  while (currentStart.isBefore(end) || currentStart.isSame(end, 'day')) {
    let currentEnd;
    
    if (granularity === TimeGranularity.CUSTOM) {
      // For custom interval, add the specified number of days (minus 1 to make it inclusive)
      currentEnd = currentStart.add(interval - 1, 'day');
      
      // If currentEnd goes beyond the overall end date, cap it at the end date
      if (currentEnd.isAfter(end)) {
        currentEnd = end;
      }
    } else {
      // For standard intervals, use endOf to get the proper period end
      currentEnd = currentStart.endOf(unit as any);
      
      // If currentEnd goes beyond the overall end date, cap it at the end date
      if (currentEnd.isAfter(end)) {
        currentEnd = end;
      }
    }
    
    // Format the label based on the granularity
    let label = '';
    switch (granularity) {
      case TimeGranularity.DAILY:
        label = currentStart.format('MMM DD');
        break;
      case TimeGranularity.WEEKLY:
        label = `${currentStart.format('MMM DD')} - ${currentEnd.format('MMM DD')}`;
        break;
      case TimeGranularity.MONTHLY:
        label = currentStart.format('MMM YYYY');
        break;
      case TimeGranularity.CUSTOM:
        label = `${currentStart.format('MMM DD')} - ${currentEnd.format('MMM DD')}`;
        break;
    }
    
    ranges.push({
      start: currentStart.format(DATE_FORMATS.YYYY_MM_DD),
      end: currentEnd.format(DATE_FORMATS.YYYY_MM_DD),
      label
    });
    
    // Move to the next period
    currentStart = currentStart.add(interval, unit as any);
    
    // Check if we've gone past the end date
    if (granularity !== TimeGranularity.CUSTOM && currentStart.isAfter(end)) {
      break;
    }
  }
  
  return ranges;
};